docs/shim-rune.md

```diff
@@ -0,0 +1,241 @@
+communication protocol is referd to
+[kata containers](https://github.com/confidential-containers/kata-containers/blob/CCv0/src/libs/protocols/protos/image.proto).
+
+```proto
+syntax = "proto3";
+...
+// Image defines the public APIs for managing images.
+service Image {
+    // PullImage pulls an image with authentication config.
+    rpc PullImage(PullImageRequest) returns (PullImageResponse) {}
+}
+
+message PullImageRequest {
+    // Image name (e.g. docker.io/library/busybox:latest).
+    string image = 1;
+    // Unique image identifier, used to avoid duplication when unpacking the image layers.
+    string container_id = 2;
+    // Use USERNAME[:PASSWORD] for accessing the registry
+    string source_creds = 3;
+}
+
+message PullImageResponse {
+    // Reference to the image in use. For most runtimes, this should be an
+    // image ID or digest.
+    string image_ref = 1;
+}
+```
+
+## 3. Create app container
+
+Ideally, just like `kata-agent`, the agent enclave container generates the bundle of the app
+container based on occlum or gramine. In this way, shim-rune can avoid generating the app container
+bundle which depends on specific Libos (occlum or gramine).
+
+But LibOS has limited capabilities such as:
+
+- For UnionFS supported by Occlum, the lower layer must be a single-layer FUSE encrypted file
+system, and the app container image to be aggregated by agent enclave may contain multiple layers,
+that is, the lower layer is not a single layer.
+- Gramine may not be able to easily support the ability to aggregate filesystems.
+
+Possible solutions:
+
+- In the long run, promote LibOS to realize the aggregation capability of supporting multiple
+lower layers.
+- In the short to medium term, the enclave-agent program cannot rely on the rootfs aggregation
+capability of a specific LibOS to construct the oci bundle of the app container. shim-rune is
+responsible for generating the bundle of the app container.
+
+Considering these limitations and after discussing with Libos team, the first stage enclave-cc
+uses the following methods to run the app enclave container.
+
+### Occlum
+
+Refer to the occlum guide of
+[Runtime boot pre-generated UnionFS image](https://github.com/qzheng527/occlum/blob/combine_jsons/demos/runtime_boot).
+A pre-created oci bundle for the boot instance is needed in the host. The boot instance is
+responsible for using the customized `init`, mount and boot a pre-generated UnionFS image.
+
+Considering the implementation of occlum dynamic mount image, shim-rune should overlay the
+`boot instance bundle` and `encrypted unionfs image` to generate the final bundle of the app
+container. For example:
+
+```sh
+mount -t overlay overlay -o
+  lowerdir= <path of encrypted unionfs image>:<path of boot instance bundle>,
+  upperdir=upper, workdir=work merged
+```
+
+Then the question is where is the location of `boot instance bundle` and `encrypted unionfs image`?
```

Comment:
Hi, please let me explain how the agent enclave works first so that we can reach a consensus.
Due to the limitation of Occlum Libos, the implementation of agent-enclave is as follows:

1. When the shim-rune sends the `image_id` to the agent enclave container, the agent enclave container uses [image-rs](https://github.com/confidential-containers/image-rs) to pull the app image, and `image-rs` will unpack the N compressed layers into N directories under the `CC_IMAGE_WORK_DIR` directory.

2. The agent enclave program will first mount the fixed host directory into the enclave through the Occlum `mount` function. Please refer to [link](https://github.com/zhiwei-intel-h/image-rs/blob/dev-occlum-adaption/src/snapshots/unionfs.rs#L107).

3. The agent enclave program copies the contents of the N layers of the image to the fixed host directory. The content combination of N layers finally generates an encrypted unionfs (based on sefs) image on the host.

Such an implementation has a big performance problem. 

A possible solution is: 

- Libos (Occlum /Gramine) supports functions similar to overlay, avoids copying multiple layers, and supports building encrypted unionfs layer(sefs based) for each layer, so as to realize layer sharing between containers. In fact, the Occlum team is working on it.

Reply:
> How the design described in this doc works when there are non-encrypted layers in the image?

I don't see this part answered. If we have a big performance problem with:

"The agent enclave program copies the contents of the N layers of the image to the fixed host directory"

why not only copy those layers that are encrypted (although it might be this does not help much if it's more common that the images have all the layers encrypted vs just a subset of them)